### PR TITLE
Merge statements

### PR DESCRIPTION
**Pull request by ravjotbrar**
### Summary

This PR finalizes the merge prototype. See linked issue below for more details. 

### Description

The changes include adding the mergeKey option which can be a list of columns, making use of the copy column list option if the column names of the target table and df are different, creating a true temporary table in Vertica, and ignoring the overwrite mode during a merge. Additionally, PR adds appropriate unit tests and integration tests. 

### Related Issue

https://github.com/vertica/spark-connector/issues/129

### Additional Reviewers

@alexr-bq 
@jonathanl-bq 
@alexey-temnikov

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -124,6 +124,7 @@ Below is a detailed list of connector options that are used in the options map:
...
+| merge_key                                      | String                                              | In order to execute a merge statement in our Spark Connector, the user needs to pass in a mergeKey, which will need to be a list of comma-separated column attributes to join the existing table and temporary table on. If this option is not populated, a write will be performed as usual without the use of a temporary table.                                                                                                                                                                                                                                                                           | No                                             |
```
Comment (alexr-bq) [inline README.md:127]: change mergeKey to merge key or merge_key

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala [hunk 2]
```diff
@@ -256,43 +285,63 @@ class VerticaDistributedFilesystemWritePipe(val config: DistributedFilesystemWri
...
-        columnList,
-        url,
-        rejectsTableName,
-        "parquet"
-      )
+      fullTableName <- if(config.mergeKey.isDefined) Right(tempTableName.getFullTableName) else Right(config.tablename.getFullTableName)
 
-      rowsCopied <- performCopy(copyStatement, config.tablename).left.map(_.context("commit: Failed to copy rows"))
+      copyStatement <- if(config.mergeKey.isDefined) {
+          Right(buildCopyStatement(fullTableName.toString(), "", url, rejectsTableName, "parquet"))
```
Comment (jonathanl-bq) [inline connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:318]: The empty string parameter is unclear to me. I think it might be better off as a default argument or maybe this should use named parameters.

Comment (jonathanl-bq) [reply connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:318]: Actually, why do we build a copy statement with no columns listed? What does that do?

File: examples/merge-example/src/main/scala/example/TestUtils.scala [hunk 3]
```diff
@@ -0,0 +1,110 @@
+// (c) Copyright [2020-2021] Micro Focus or one of its affiliates.
```
Comment (jonathanl-bq) [inline examples/merge-example/src/main/scala/example/TestUtils.scala:1]: I'm not a huge fan of how much we've been copying these TestUtils. They usually provide way more functionality than we really need. We might want to consider removing these modules and just inlining the utility functions used.

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala (hunk 2 above)
Comment (alexr-bq) [inline connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:318]: Could move the if statement to inside the function call for less repetition.

File: functional-tests/src/main/scala/com/vertica/spark/functests/EndToEndTests.scala [hunk 4]
```diff
@@ -3310,4 +3310,83 @@ class EndToEndTests(readOpts: Map[String, String], writeOpts: Map[String, String
     assert ( rowsLoaded == numDfRows )
     TestUtils.dropTable(conn, tableName)
   }
+
```
Comment (jonathanl-bq) [inline functional-tests/src/main/scala/com/vertica/spark/functests/EndToEndTests.scala:3313]: I'd like to see test cases for:
- merge with no existing table in Vertica
- some updates and some inserts
- all inserts
- all updates
- no updates or inserts

**Review by jonathanl-bq — CHANGES_REQUESTED** [review]

File: functional-tests/src/main/scala/com/vertica/spark/functests/EndToEndTests.scala [hunk 5]
```diff
@@ -3310,4 +3310,83 @@ class EndToEndTests(readOpts: Map[String, String], writeOpts: Map[String, String
...
+  it should "Merge with existing table in Vertica" in {
```
Comment (alexr-bq) [inline functional-tests/src/main/scala/com/vertica/spark/functests/EndToEndTests.scala:3314]: Add test case for multiple columns in merge key

**Review by alexr-bq — CHANGES_REQUESTED** [review]

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala [hunk 6]
```diff
@@ -256,43 +285,63 @@ class VerticaDistributedFilesystemWritePipe(val config: DistributedFilesystemWri
 
   def commit(): ConnectorResult[Unit] = {
     val globPattern: String = "*.parquet"
-
+    val tableNameMaxLength = 30
```
[the remark this answers is not in the document]
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:288]: I was playing around with the max table name length with the addition of a temp table and have initially moved the location of it. Then I realized that it doesn't impact merge statements so I put it back in an arbitrary but appropriate spot.

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala [hunk 7]
```diff
@@ -256,43 +285,63 @@ class VerticaDistributedFilesystemWritePipe(val config: DistributedFilesystemWri
...
 
       _ <- tableUtils.updateJobStatusTable(config.tablename, config.jdbcConfig.auth.user, faultToleranceResults.failedRowsPercent, config.sessionId, faultToleranceResults.success)
 
       _ <- if (faultToleranceResults.success) Right(()) else Left(FaultToleranceTestFail())
+
+      mergeStatement <- if (config.mergeKey.isDefined) {
+                          Right(buildMergeStatement(config.tablename, columnList, tempTableName.getFullTableName))
+                        }
+                        else {
+                          Right((""))
```
[the remark this answers is not in the document]
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:341]: Good catch

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala [hunk 8]
```diff
@@ -147,6 +148,35 @@ class VerticaDistributedFilesystemWritePipe(val config: DistributedFilesystemWri
...
   }
 
+  def buildMergeStatement(targetTableName: TableName, columnList: String, tempTable: String): String = {
+    val targetTable = targetTableName.getFullTableName
+    val updateColValues = schemaTools.getUpdateValues(jdbcLayer, targetTableName, tempTableName, config.copyColumnList)
+    val insertColValues = schemaTools.getInsertValues(jdbcLayer, tempTableName, config.copyColumnList)
+    val mergeList = config.mergeKey match {
+      case Some(keys) => {
+        keys.toString.split(",").toList.map(col => s"target.$col=temp.$col").mkString(" AND ")
+      }
```
[the remark this answers is not in the document]
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:158]: Good point

File: connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala [hunk 9]
```diff
@@ -87,6 +87,25 @@ trait SchemaToolsInterface {
...
   def makeColumnsString(columnDefs: Seq[ColumnDef], requiredSchema: StructType): String
+
+  /**
+   * Gets a list of column values to be inserted within a merge.
+   *
+   * @param columnDefs List of column definitions from the Vertica table.
+   * @param copyColumnList String of columns passed in by user as a configuration option.
+   * @return String of values to append to INSERT VALUES in merge.
+   */
+  def getInsertValues(jdbcLayer: JdbcLayerInterface, tableName: TableName, copyColumnList: Option[ValidColumnList]): String
```
[the remark this answers is not in the document]
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala:98]: Agreed

File: examples/merge-example/src/main/scala/example/TestUtils.scala (hunk 3 above)
Comment (ravjotbrar) [reply examples/merge-example/src/main/scala/example/TestUtils.scala:1]: That's true

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala (hunk 2 above)
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:318]: That parameter is only required for a normal write with customCopyList (I think). Otherwise it's not necessary. I'll change it to a default argument.

File: README.md (hunk 1 above)
Comment (ravjotbrar) [reply README.md:127]: Good catch

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala [hunk 10]
```diff
@@ -256,43 +285,63 @@ class VerticaDistributedFilesystemWritePipe(val config: DistributedFilesystemWri
...
-        rejectsTableName,
-        "parquet"
-      )
+      fullTableName <- if(config.mergeKey.isDefined) Right(tempTableName.getFullTableName) else Right(config.tablename.getFullTableName)
 
-      rowsCopied <- performCopy(copyStatement, config.tablename).left.map(_.context("commit: Failed to copy rows"))
+      copyStatement <- if(config.mergeKey.isDefined) {
+          Right(buildCopyStatement(fullTableName.toString(), "", url, rejectsTableName, "parquet"))
+        }
+                       else {
```
[the remark this answers is not in the document]
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:320]: Done

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala [hunk 11]
```diff
@@ -256,43 +285,63 @@ class VerticaDistributedFilesystemWritePipe(val config: DistributedFilesystemWri
...
+          Right(buildCopyStatement(fullTableName.toString(), columnList, url, rejectsTableName, "parquet"))
+        }
+
+      rowsCopied <- if (config.mergeKey.isDefined) {
+        Right(performCopy(copyStatement.toString, tempTableName).left.map(_.context("commit: Failed to copy rows into temp table")))}
+                    else {
+        Right(performCopy(copyStatement.toString, config.tablename).left.map(_.context("commit: Failed to copy rows into target table")))
+        }
```
[the remark this answers is not in the document]
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:328]: Done.

File: docker/Dockerfile [hunk 12]
```diff
@@ -6,7 +6,8 @@ ENV SBT_VERSION 1.3.13
 RUN yum install -y java-11-openjdk && \
     yum install -y epel-release && \
     yum update -y && yum install -y wget && \
-    wget http://dl.bintray.com/sbt/rpm/sbt-$SBT_VERSION.rpm && \
-    yum install -y sbt-$SBT_VERSION.rpm
+    curl -L https://www.scala-sbt.org/sbt-rpm.repo > sbt-rpm.repo && \
```
[the remark this answers is not in the document]
Comment (ravjotbrar) [reply docker/Dockerfile:9]: Yeah you're right, I'll revert it and merge from main.

File: examples/run-example.sh [hunk 13]
```diff
@@ -1,11 +1,11 @@
 if [ "$1" == "clean" ]
   then
-    wget -P ../../ https://apache.mirror.colo-serv.net/spark/spark-3.0.2/spark-3.0.2-bin-without-hadoop.tgz
+    wget -P ../../ https://mirror.dsrg.utoronto.ca/apache/spark/spark-3.1.2/spark-3.1.2-bin-without-hadoop.tgz
```
[the remark this answers is not in the document]
Comment (ravjotbrar) [reply examples/run-example.sh:3]: Sounds good.

File: functional-tests/src/main/scala/com/vertica/spark/functests/EndToEndTests.scala (hunk 5 above)
Comment (ravjotbrar) [reply functional-tests/src/main/scala/com/vertica/spark/functests/EndToEndTests.scala:3314]: Done

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala (hunk 2 above)
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala]: Done

File: functional-tests/src/main/scala/com/vertica/spark/functests/EndToEndTests.scala (hunk 4 above)
Comment (ravjotbrar) [reply functional-tests/src/main/scala/com/vertica/spark/functests/EndToEndTests.scala:3313]: Added all test cases except the last, as that doesn't seem possible.

File: examples/basic-write/build.sbt [hunk 14]
```diff
@@ -12,7 +12,7 @@
 // limitations under the License.
 
 scalaVersion := "2.12.12"
-name := "spark-vertica-connector-basic-write-example"
```
[the remark this answers is not in the document]
Comment (ravjotbrar) [reply examples/basic-write/build.sbt]: Changed

File: README.md [hunk 15]
```diff
@@ -124,6 +124,7 @@ Below is a detailed list of connector options that are used in the options map:
...
+| merge_key                                      | String                                              | In order to execute a merge statement in our Spark Connector, the user needs to pass in a merge key, which will be a list of comma-separated column attributes to join the existing table and temporary table on. Note: the merge key must be unique in the dataframe for the merge to work. Additionally, while you may use copy_column_list to get around differing column names between the dataframe and existing table, the column names of the merge_key must match in both schemas.                                                                                                           |                                                                                                                                                                 | No                                             |
```
Comment (alexey-temnikov) [inline README.md:127]: Should we add small paragraph explaining how Merge statements are supported? Diagram from https://github.com/vertica/spark-connector/issues/129 might be very helpful.

File: connector/src/main/scala/com/vertica/spark/datasource/core/DSConfigSetup.scala [hunk 16]
```diff
@@ -362,6 +362,13 @@ object DSConfigSetupUtils {
     }
   }
 
+  def getMergeKey(config: Map[String, String]) : ValidationResult[Option[ValidColumnList]] = {
+    config.get("merge_key")match {
```
Comment (alexey-temnikov) [inline connector/src/main/scala/com/vertica/spark/datasource/core/DSConfigSetup.scala:366]: nit: space before match?

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala [hunk 17]
```diff
@@ -45,6 +45,7 @@ class VerticaDistributedFilesystemWritePipe(val config: DistributedFilesystemWri
           extends VerticaPipeInterface with VerticaPipeWriteInterface {
 
   private val logger = LogProvider.getLogger(classOf[VerticaDistributedFilesystemWritePipe])
+  private val tempTableName = new TableName(config.tablename.name + "_" + config.sessionId, None)
```
Comment (alexey-temnikov) [inline connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:48]: nit: Do we need "new" modifier here?

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala [hunk 18]
```diff
@@ -97,15 +98,15 @@ class VerticaDistributedFilesystemWritePipe(val config: DistributedFilesystemWri
       _ = setSparkCalendarConf()
 
       // If overwrite mode, remove table and force creation of new one before writing
-      _ <- if(config.isOverwrite) tableUtils.dropTable(config.tablename) else Right(())
+      _ <- if(config.isOverwrite && !config.mergeKey.isDefined) tableUtils.dropTable(config.tablename) else Right(())
```
Comment (alexey-temnikov) [inline connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:101]: nit: should we use `config.mergeKey.isEmpty` instead of `!config.mergeKey.isDefined`?

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala [hunk 19]
```diff
@@ -97,15 +98,15 @@ class VerticaDistributedFilesystemWritePipe(val config: DistributedFilesystemWri
...
 
       // Create the table if it doesn't exist
       tableExistsPre <- tableUtils.tableExists(config.tablename)
 
       // Overwrite safety check
-      _ <- if (config.isOverwrite && tableExistsPre) Left(DropTableError()) else Right(())
+      _ <- if (config.isOverwrite && !config.mergeKey.isDefined && tableExistsPre) Left(DropTableError()) else Right(())
```
Comment (alexey-temnikov) [inline connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:107]: nit: same as above, isEmpty?

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala [hunk 20]
```diff
@@ -256,43 +294,59 @@ class VerticaDistributedFilesystemWritePipe(val config: DistributedFilesystemWri
...
-      copyStatement = buildCopyStatement(config.tablename.getFullTableName,
-        columnList,
-        url,
-        rejectsTableName,
-        "parquet"
-      )
+      fullTableName <- if(config.mergeKey.isDefined) Right(tempTableName.getFullTableName) else Right(config.tablename.getFullTableName)
 
-      rowsCopied <- performCopy(copyStatement, config.tablename).left.map(_.context("commit: Failed to copy rows"))
+      copyStatement = buildCopyStatement(fullTableName.toString(), columnList, url, rejectsTableName, "parquet")
```
Comment (alexey-temnikov) [inline connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:326]: nit: is toString() needed?

File: connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala [hunk 21]
```diff
@@ -87,6 +87,25 @@ trait SchemaToolsInterface {
    * @return List of columns in matches.
    */
   def makeColumnsString(columnDefs: Seq[ColumnDef], requiredSchema: StructType): String
+
+  /**
+   * Gets a list of column values to be inserted within a merge.
+   *
+   * @param columnDefs List of column definitions from the Vertica table.
```
Comment (alexey-temnikov) [inline connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala:94]: What is columnDefs, it is not used in the method?

File: connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala [hunk 22]
```diff
@@ -87,6 +87,25 @@ trait SchemaToolsInterface {
...
+   * @param copyColumnList String of columns passed in by user as a configuration option.
+   * @return String of values to append to INSERT VALUES in merge.
+   */
+  def getMergeInsertValues(jdbcLayer: JdbcLayerInterface, tableName: TableName, copyColumnList: Option[ValidColumnList]): ConnectorResult[String]
+
+  /**
+   * Gets a list of column values and their updates to be updated within a merge.
+   *
+   * @param columnDefs List of column definitions from the Vertica table.
```
Comment (alexey-temnikov) [inline connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala:103]: What is columnDefs, it is not used in the method?

File: connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala [hunk 23]
```diff
@@ -320,5 +337,37 @@ class SchemaTools extends SchemaToolsInterface {
...
+    val valueList = getColumnInfo(jdbcLayer, tableName) match {
+      case Right(info) => Right(info.map(x => "temp." + x.label ).mkString(","))
+      case Left(err) => Left(JdbcSchemaError(err))
+    }
+    valueList
+  }
+
+  def getMergeUpdateValues(jdbcLayer: JdbcLayerInterface, tableName: TableName, tempTableName: TableName, copyColumnList: Option[ValidColumnList]): ConnectorResult[String] = {
+    val columnList = copyColumnList match {
+      case Some(list) => {
```
Comment (alexey-temnikov) [inline connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala:351]: nit: Redundant braces in case clause.

File: connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala [hunk 24]
```diff
@@ -320,5 +337,37 @@ class SchemaTools extends SchemaToolsInterface {
...
+        val customColList = list.toString.split(",").toList
+        val colList = getColumnInfo(jdbcLayer, tempTableName) match {
+          case Right(info) => {
```
Comment (alexey-temnikov) [inline connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala:354]: nit: Redundant braces in case clause.

File: connector/src/test/scala/com/vertica/spark/datasource/core/DSWriterTest.scala [hunk 25]
```diff
@@ -41,7 +41,8 @@ class DSWriterTest extends AnyFlatSpec with BeforeAndAfterAll with MockFactory {
     copyColumnList = None,
     sessionId = "id",
     failedRowPercentTolerance =  0.0f,
-    filePermissions = ValidFilePermissions("777").getOrElse(throw new Exception("File perm error")))
+    filePermissions = ValidFilePermissions("777").getOrElse(throw new Exception("File perm error")),
+    mergeKey = None)
```
Comment (alexey-temnikov) [inline connector/src/test/scala/com/vertica/spark/datasource/core/DSWriterTest.scala:45]: Should we use default value in DistributedFilesystemWriteConfig for mergeKey as None (so APIs remain compatible?)

File: README.md (hunk 15 above)
Comment (jonathanl-bq) [inline README.md:127]: An example comma separated string would be nice. Is it "foo, bar, baz"? Or is it "foo,bar,baz"? Or both?

**Review by jonathanl-bq — CHANGES_REQUESTED** [review]

File: connector/src/test/scala/com/vertica/spark/datasource/core/DSWriterTest.scala (hunk 25 above)
Comment (ravjotbrar) [reply connector/src/test/scala/com/vertica/spark/datasource/core/DSWriterTest.scala:45]: Sure, I don't see an issue with that.

File: connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala (hunk 21 above)
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala:94]: Good catch

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala (hunk 20 above)
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:326]: I think it was before, but I might've made some changes that make it unnecessary. Changed.

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala (hunk 18 above)
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:101]: That sounds like it would be more readable.

File: README.md (hunk 15 above)
Comment (ravjotbrar) [reply README.md:127]: Where do you think the best place to add it is? I think it'll make the description of the option too wordy if we put it here.

File: connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala (hunk 17 above)
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/datasource/core/VerticaDistributedFilesystemWritePipe.scala:48]: I guess not, since it's a case class. Removed.

File: README.md (hunk 15 above)
Comment (alexey-temnikov) [reply README.md]: Yes, agree. Considering we have a dedicated example app for merge statements, I'm ok just with this description.

File: connector/src/main/scala/com/vertica/spark/config/WriteConfig.scala [hunk 26]
```diff
@@ -82,7 +83,8 @@ final case class DistributedFilesystemWriteConfig(jdbcConfig: JDBCConfig,
                                                   copyColumnList: Option[ValidColumnList],
                                                   sessionId: String,
                                                   failedRowPercentTolerance: Float,
-                                                  filePermissions: ValidFilePermissions
+                                                  filePermissions: ValidFilePermissions,
+                                                  mergeKey: Option[ValidColumnList]= None
```
Comment (jonathanl-bq) [inline connector/src/main/scala/com/vertica/spark/config/WriteConfig.scala:87]: nit: space before equals

File: connector/src/main/scala/com/vertica/spark/util/table/TableUtils.scala [hunk 27]
```diff
@@ -56,6 +66,15 @@ trait TableUtilsInterface {
...
 
+  /**
+   * Creates a temporary table.
+   *
+   * @param tablename Name of table
+   * @param schema Spark schema of data we want to write to the table
+   * @param strlen Length to use for strings in Vertica string types
+   */
+  def createTempTable(tablename: TableName, schema: StructType, strlen: Long): ConnectorResult[Unit]
+
```
Comment (jonathanl-bq) [inline connector/src/main/scala/com/vertica/spark/util/table/TableUtils.scala:77]: Can this not just be a private method? As far as I can tell, it's just a helper. There's no need to expose this to the implementations.

File: connector/src/main/scala/com/vertica/spark/util/table/TableUtils.scala [hunk 28]
```diff
@@ -210,6 +233,12 @@ class TableUtils(schemaTools: SchemaToolsInterface, jdbcLayer: JdbcLayerInterfac
       .left.map(err => err.context("JDBC Error dropping table"))
   }
 
+  override def createTempTable(tablename: TableName, schema: StructType, strlen: Long): ConnectorResult[Unit] = {
+    val statement= buildCreateTableStmt(tablename, schema, strlen, true)
```
Comment (jonathanl-bq) [inline connector/src/main/scala/com/vertica/spark/util/table/TableUtils.scala:237]: nit: space before equals sign

**Review by jonathanl-bq — CHANGES_REQUESTED** [review]

File: connector/src/main/scala/com/vertica/spark/util/table/TableUtils.scala (hunk 27 above)
Comment (ravjotbrar) [reply connector/src/main/scala/com/vertica/spark/util/table/TableUtils.scala:77]: You mean the buildCreateTableStmt method, right?

Comment (jonathanl-bq) [reply connector/src/main/scala/com/vertica/spark/util/table/TableUtils.scala:67]: yeah, that's what I meant.

**Review by jonathanl-bq — APPROVED** [review]

**Review by alexey-temnikov — APPROVED** [review]

**Review by alexey-temnikov — APPROVED** [review]
